### DISCUSSION TITLE
Updating JSON fields with another JSON structure, merges with existing structure instead of replacing the old one

### DISCUSSION THREAD

**Discussion by errol59** [mikro-orm/mikro-orm#5410] [General]
### Describe the bug

In the [docs](https://mikro-orm.io/docs/entity-helper#updating-entity-values-with-entityassign) the following statement is made:

> By default, entity.assign(data) behaves similar to Object.assign(entity, data)...

However, that statement holds not entirely true if you look at my reproduction example.

### Reproduction

```
// table structure: `name` has column type `varchar(255)` and `structure` has column type `jsonb`

const person = new Author('me', { foo: 1, bar: 2  });
// { name: 'me', structure: { foo: 1, bar: 2 } }

wrap(person).assign({ structure: { bar: 3, baz: 4 } });
// Expected: { name: 'me', structure: { bar: 3, baz: 4 } }
// Reality: { name: 'me', structure: { foo: 1, bar: 3, baz: 4 } }
```

```
const x = { name: 'me', structure: { foo: 1, bar: 2 } }
console.log(Object.assign(x, { structure: { bar: 3, baz: 4 }}))
// { name: 'me', 'structure': { 'bar': 3, 'baz': 4 } }
```

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.1.4

### Node.js version

20.10.0

### Operating system

Docker WSL

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
How does your entity look like? There are tests for this behavior, it really does work this way by default - you would need to `mergeObjectProperties` option for the objects to merge.

Note that this will work only for POJOs, maybe you use a custom class for the data you are assigning?

> 6.1.4

And be sure to update to latest version first, that should be always your first step when something looks weird.

**Reply by errol59** [reply]
> And be sure to update to latest version first, that should be always your first step when something looks weird.

I just updated my package to the latest version, but this issue still persists

> ...There are tests for this behavior, it really does work this way by default - you would need to mergeObjectProperties option for the objects to merge.

My reproduction example is not really representative of what I try to do. Below a more exact example

```
const person = (await MikroORM.init(connectionOptions)).em.getRepository(Model).findOne({ name: 'me' });
// Person { name: 'me', structure: { foo: 1, bar: 2 } }

wrap(person).assign({ structure: { bar: 3, baz: 4 } });
// Expected: Person { name: 'me', structure: { bar: 3, baz: 4 } }
// Reality: Person { name: 'me', structure: { foo: 1, bar: 3, baz: 4 } }
```

Instead of creating an entity instance, I'm retrieving an existing one from the database and try to update that one with the `wrap.assign` method. The property `mergeObjectProperties` is not being used.

>  How does your entity look like?

Below a really concise and obscured part of my Entity. All other basic columns are stripped away.

```
@Entity({ tableName: 'workflow_versions' })
export class Person extends BaseEntity {
    @ManyToOne({
        entity: () => Organization,
        inversedBy: (organization) => organization.people,
        deleteRule: 'cascade',
        ref: true,
    })
    organization!: Ref<Organization>;

    @OneToMany({
        entity: () => Job,
        mappedBy: (jobs) => jobs.person,
        orphanRemoval: false,
    })
    job = new Collection<Job>(this);

    @Property()
    name!: string;

    @Property({ columnType: 'jsonb', nullable: true })
    structure!: Pick<Structure, 'nodes' | 'connections'> | null;
}
```

**Reply by B4nan** [reply]
Uh, now I'm confused, the `mergeObjectProperties` apparently defaults to true ([since v5](https://mikro-orm.io/docs/upgrading-v4-to-v5#changes-in-assign-helper)).

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/entity/EntityAssigner.ts#L49

So the docs are wrong.

**Reply by B4nan** [reply]
I am starting to think that the flag is used for two different things (JSON props and embedded props), and we should in fact have two flags, and only the one for embedded props should be enabled by default.

**Reply by errol59** [reply]
> I am starting to think that the flag is used for two different things (JSON props and embedded props), and we should in fact have two flags, and only the one for embedded props should be enabled by default.

I was thinking the same. The Docs mention embedded props/database relations, but doesn't mention anything regards normal JSON values that can exist inside a single database column.

Also, explicitly setting `mergeObjectProperties` to false solved my issue.

**Reply by B4nan** [reply]
The change was introduced here: 7b7c3a22fe517e13a1a610f142c59e758acd3c3f, so the motivation was indeed embedded props.

**Reply by B4nan** [reply]
I guess I will give this a go, it feels a bit breaking, but since the docs still mention the old default, it feels more like a fix (and the default behavior for JSON props really feels like a bug, I completely agree with that).

**Reply by B4nan** [reply]
bc9f6f5bbecac302f0beb227703a9654f71d967b
